src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+
+namespace TorchSharp
+{
+    /// <summary>
+    /// Manages dispose scopes, that can make automatic tensor disposal easier. Note that the
+    /// DisposeManager is thread local. The DisposeScopeManager can also manage other disposables, such as training
+    /// batches and the like.
+    /// </summary>
+    public class DisposeScopeManager
+    {
+        private int _createdCount;
+        private int _disposedCount;
+
+        static DisposeScopeManager()
+        {
+            torch.Tensor.OnTensorCreateCallback.Add(TryRegisterInDisposeScope);
+            torch.Tensor.OnTensorDisposeCallback.Add(TryUnregisterInDisposeScope);
+        }
+
+        [ThreadStatic] private static DisposeScopeManager _singleton;
+
+        public static DisposeScopeManager Singleton = (_singleton ??= new DisposeScopeManager());
+
+        internal Stack<DisposeScope> DisposeScopeStack { get; } = new();
+        internal IDisposable CurrentlyDisposing { get; set; }
+
+        public int CreatedCount => _createdCount;
+        public int DisposedCount => _disposedCount;
+        public int TotalCount => _createdCount - _disposedCount;
+
+        public static long TotalAllocatedSize => Singleton.InnerTotalAllocatedSize;
```

Comment:
This is misleading, because it does not return total allocated for all threads, but the name does not have any indication of that. Why not just keep it an instance property?

Reply:
I removed allocated size, @NiklasGustafsson has mentioned that it's too approximative anyway.

I made the thread "singleton" private, so I had to keep the counts public, but I did this;

        /// <summary>
        /// The number of disposables that are currently live on the current thread. It's aproximate, see
        /// Tensor.TotalCount.
        /// </summary>
        public static longThreadTotalLiveCount => ThreadSingleton._createdCount - ThreadSingleton._disposedCount;

Lemme know  what you think